Companion Mode + New Flow

artosimonyan opened the pull request:
- Companion mode returns to Koolo
- Completely revamped creating/joining games logic
- Greatly improved online check logic and eliminated a lot of edge cases
- Improved failure to select Character with maximum attempts and a client restart.

hectorgimenez commented on internal/event/events.go:
```diff
@@ -151,3 +149,33 @@ func GamePaused(be BaseEvent, paused bool) GamePausedEvent {
+}
+
+// ResetCompanionGameInfoEvent is sent when the Leader finishes a game, preventing the companions from joining it
+func ResetCompanionGameInfo(be BaseEvent, leader string) ResetCompanionGameInfoEvent {
```
Maybe just call it GameFinished or whatever? the responsibility of reseting the companion game info is based on the handler, maybe the GameFinished event can be used for other stuff too.

artosimonyan replied:
We already have GameFinished event which we use for other stats. The idea here is to separate them with other events that we can use, even outside of the bot (ea discord integrations etc), specifically for companion to allow for more customization and actions.

davidfvsilva commented:
The Leader seems to be idling in Lobby after X amount of runs. It can vary. I've been testing this and have seen nothing in logs, so it must be silent errors or an edge case.  It can take 1 hour, it can take 8 hours, but eventually it stops. More logs through the process could help.

davidfvsilva commented:
After testing it for a few days with groups, there is an issue where the Leader stops creating games (it usually doesn't take much time, can be less than an hour, but it can take more time). You can also have cases where the companion doesn't get the join request (it will skip a run and join the following one), and where a companion can get a join request for a previous game (it will try to join an old game name).

Antizerg commented:
I also encountered the issue where the leader stops creating games, but this was not a new bug for me so hard to say if the companion feature created the issue or not.  I switched to not creating lobby games to solve this in the past, so I never went deeper with it.

artosimonyan commented:
So ... it's just a problem when the leader creates Lobby games? If so, just create normal ones. And by stops do you mean he does nothing (idles for 10+ minutes) or what? 

I really need you to provide more information here and not vague explanations

davidfvsilva commented:
> So ... it's just a problem when the leader creates Lobby games? If so, just create normal ones. And by stops do you mean he does nothing (idles for 10+ minutes) or what? 
> 
> I really need you to provide more information here and not vague explanations

it will idle indefinitely in lobby like I mentioned before. Unfortunately, there isn't anything noticeable in the logs.

artosimonyan commented:
@Antizerg @davidfvsilva, test pressing escape to get back to the character selection screen and let me know if that will make it start doing stuff again. 
Preferably if you can report back before tomorrow's reset, would be wonderful.

davidfvsilva commented:
> @Antizerg @davidfvsilva, test pressing escape to get back to the character selection screen and let me know if that will make it start doing stuff again. Preferably if you can report back before tomorrow's reset, would be wonderful.

you need to manually enter a game to get it running again. from a distance you can also, stop-start the bot with a discord command

RobotGaston commented:
### **[companion receiving wrong game number from the leader]**

thank you for this excellent feature.

there is a problem with leader-companion communication unfortunately.

it is happening occasionally NOT all the time :(

leader is running **correctly** but _companion_ can receive "incorrect game name" when the next game is created by a leader, which makes companion to join previous game, and not the new game where the leader is.

in a log it looks like this:

level=INFO msg="Companion join game event received" supervisor=xxx leader=xx name=xx0 password=x
where "name=xx0" xx0 - number "0" is incorrect game number. In reality for leader the game name is "xx1" but companion received 0 instead 1 in a games name...:(

-                                            -                                  -                                     -
and also I observed that sometimes companion is receiving an empty game name+password in a log...
level=INFO msg="Companion join game event received" supervisor="" leader="" name="" password=""

also happens randomly, leader and companion can run 45+ games together without a problem and in 46 it can occur for no any reason at all..

something might be going on In game name+password communication in a companion's part of the program...
leader is counting up and creating new "higher number" games fine.

could this be looked into please and may be corrected?
Thank you.

artosimonyan commented:
The issue doesn't seem directly connected with the companion but with the out of game loop. I'll have to dig into it ... not sure how exactly but this needs work

artosimonyan commented:
@davidfvsilva, @Antizerg Updated the flow, added some more checks and added debug info via the SetLastAction. Please test again and let me know if there are any issues.

Antizerg commented:
> @davidfvsilva, @Antizerg Updated the flow, added some more checks and added debug info via the SetLastAction. Please test again and let me know if there are any issues.

no issues after a 6 hour run

Neo-Silver commented on internal/bot/companion.go:
```diff
@@ -0,0 +1,60 @@
+			// Check if the leader matches the one in our config or no leader set.
+			// Additional check for if LeaderName is the same as the character name for Manual join triggers
+			if h.cfg.Companion.LeaderName == "" || evt.Leader == h.cfg.Companion.LeaderName || h.cfg.CharacterName == evt.Leader {
+				h.log.Info("Companion reset game info event received", slog.String("supervisor", h.supervisor), slog.String("leader", evt.Leader))
```
I could not test it indepth, but maybe add an option for companions to leave game as soon as the leader sends the new game.

if h.manager.InGame() {
	if err := h.manager.ExitGame(); err != nil {
			}
	}
or something like this?

Antizerg commented on internal/bot/companion.go:
```diff
@@ -0,0 +1,60 @@
+			// Additional check for if LeaderName is the same as the character name for Manual join triggers
+			if h.cfg.Companion.LeaderName == "" || evt.Leader == h.cfg.Companion.LeaderName || h.cfg.CharacterName == evt.Leader {
+				h.log.Info("Companion reset game info event received", slog.String("supervisor", h.supervisor), slog.String("leader", evt.Leader))
+				h.cfg.Companion.CompanionGameName = ""
```
> I could not test it indepth, but maybe add an option for companions to leave game as soon as the leader sends the new game.
> 
> After recieving new game event: if h.manager.InGame() { if err := h.manager.ExitGame(); err != nil { } } or something like this?

this would help with just letting a companion join a game with no missions, which I feel will get a lot of the use

Neo-Silver replied:
> > I could not test it indepth, but maybe add an option for companions to leave game as soon as the leader sends the new game.
> > After recieving new game event: if h.manager.InGame() { if err := h.manager.ExitGame(); err != nil { } } or something like this?
> 
> this would help with just letting a companion join a game with no missions, which I feel will get a lot of the use

Null and I were talking about that.
It is in the work.

Neo-Silver commented on internal/bot/supervisor.go:
```diff
@@ -126,23 +127,47 @@ func (s *baseSupervisor) waitUntilCharacterSelectionScreen() error {
 				return nil
 			}
 
 			s.bot.ctx.HID.PressKey(win.VK_DOWN)
```
Why this?
It seems to move down from the in a loop in the char select screen.
Ending with the bottem most character selected. (atleast it did that for me)

Neo-Silver reviewed (commented):
Nvm - seemed to be an connection issue.

davidfvsilva commented:
Still seeing the leader stop making games after a few runs. As for having the companion quit the game, you can send an event that runs the ExitGame method on companions when the runs are done, then give them a few seconds to leave the game. Companions sometimes stop joining games, but not nearly as often as the leader stops making them. Also, you have these wierd cases where a new game event is fired from nowhere (as if it had it in queue somewhere), and theses cases were the companion is trying to join the wrong game.

RobotGaston commented:
after a 12+ hours of testing I also can confirm that, leader and companion **still** can get our of sync...

problem remains...:(

this is working much, **much** , better now, companions and leader can run more longer games streak together **without** getting out of sync now (70-80 games in a row, hours of communicating) but still...

companions keep on joining previous game with time, and not the new game where the leader is.

The game number is **off by 1** for a companions , i. e.

**leader** is in _"mygame2"_

**companions** joined _"mygame1"_ (already cleaned and played game). a previous game.

this state is not getting resolved until both koolo+game windows (mandatory) are getting full complete restart.
something is _jamming_ in game creation for the **companions** windows...

RobotGaston commented:
with these changes - [4aea0c8](https://github.com/hectorgimenez/koolo/pull/716/commits/4aea0c88a1f64aac72a4828b6444bb5b212b521a)

on my end **Leader** creating game and aborting immediately, getting into loop - spamming create game task...after 1st successful run.

foolshopedemocrazy commented:
I tried to get the fix working but I failed. Could someone kindly upload and share the compiled build version?
